src/server/lib/okta/authentication.ts

```diff
@@ -0,0 +1,165 @@
+interface AuthenticationLinks {
+  next?: HalLink;
+  prev?: HalLink;
+  cancel?: HalLink;
+  skip?: HalLink;
+  resend?: HalLink;
+}
+// https://developer.okta.com/docs/reference/api/authn/#user-object
+interface AuthenticationEmbeddedUser {
+  id: string;
+  passwordChanged?: string;
+  profile: {
+    firstName: string;
+    lastName: string;
+    locale?: string;
+    login: string;
+    timeZone?: string;
+  };
+  recovery_question?: {
+    question: string;
+  };
+}
+
+// https://developer.okta.com/docs/reference/api/authn/#authentication-transaction-object
+interface AuthenticationTransaction {
+  _embedded?: AuthenticationEmbeddedUser;
+  _links?: AuthenticationLinks;
+  expiresAt?: string;
+  factorResult?: AuthenticationFactorResult;
+  sessionToken?: string;
+  stateToken?: string;
+  status?: AuthenticationTransactionState;
+}
+
+// https://developer.okta.com/docs/reference/api/authn/#request-parameters-for-primary-authentication
+interface AuthenticationRequestParameters {
+  audience?: string;
+  context?: {
+    deviceToken?: string;
+  };
+  options?: {
+    multiOptionalFactorEnroll?: boolean;
+    warnBeforePasswordExpired?: boolean;
+  };
+  password?: string;
+  token?: string;
+  username?: string;
+}
+
+// https://developer.okta.com/docs/reference/error-codes/
+export interface OktaError {
+  errorCode: string;
+  errorSummary: string;
+  errorLink: string;
+  errorId: string;
+  errorCauses: [string?];
+}
+
+export interface FetchOktaError {
+  status: number;
+  statusText: string;
+  oktaError: OktaError;
+}
+
+const { oktaDomain } = getConfiguration();
+
+const handleResponseFailure = async (response: Response) => {
+  const parsed = (await response.json()) as OktaError;
+  switch (parsed.errorCode) {
+    // errors from https://developer.okta.com/docs/reference/api/authn/#response-parameters
+    case 'E0000004':
+      throw {
+        status: response.status,
+        statusText: response.statusText,
+        message: OktaAuthenticateErrors.AUTHENTICATION_FAILED,
+      };
+    case 'E0000047':
+      throw {
+        status: response.status,
+        statusText: response.statusText,
+        message: SignInErrors.GENERIC,
+      };
+    default:
+      throw {
+        status: response.status,
+        statusText: response.statusText,
+        message: SignInErrors.GENERIC,
+      };
+  }
+};
+
+const handleResponseSuccess = async (
+  response: Response,
+): Promise<AuthenticationTransaction> => {
+  try {
+    return (await response.json()) as AuthenticationTransaction;
+  } catch (error) {
+    throw new Error(`Error decoding JSON response: ${error}`);
+  }
+};
+
+// okta authentication endpoint
+// gateway (profile) is set up as a public application as it is
+// accessible by anyone
+// https://developer.okta.com/docs/reference/api/authn/#request-example-for-primary-authentication-with-public-application
+export const authenticate = async (email: string, password: string) => {
```

Comment:
I wonder if there's benefit in calling this file `authenticate.ts`? I subscribe to the 'one thing per file' way of dong things - where a file has one main export which has the same name as the file - which could work here? The down side to an open bucket like this is it can be very inviting and can lead to lots of things being added, reducing discoverability.

It's true though that the one file per export thing doesn't always work, especially when there's lots of shared code. I also subscribe to the 'move stuff around until it feels right' way of doing things as well!

Reply:
Sounds good to me!